lib/win-exception-handler/exception_handling/exception_handling.c

```diff
@@ -0,0 +1,71 @@
+#include <windows.h>
+#include <setjmp.h>
+#include "exception_handling.h"
+
+#define CALL_FIRST 1
+
+__declspec(thread) jmp_buf jmpBuf;
+__declspec(thread) PVOID caughtExceptionAddress;
+__declspec(thread) DWORD64 caughtInstructionPointer;
+__declspec(thread) PVOID savedStackPointer;
+__declspec(thread) BOOL exceptionHandlerInstalled = FALSE;
+__declspec(thread) BOOL alreadyHandlingException = FALSE;
+
+void longjmpOutOfHere() {
+    longjmp(jmpBuf, 1);
+}
+
+/// Get the current address that we use to jmp, the no inline is important
+static __declspec(noinline) void *get_callee_frame_address(void) {
+    return _AddressOfReturnAddress();
+}
+
+static LONG WINAPI
+exceptionHandler(struct _EXCEPTION_POINTERS *ExceptionInfo) {
+    EXCEPTION_RECORD* pExceptionRecord = ExceptionInfo->ExceptionRecord;
+    PCONTEXT pCONTEXT = ExceptionInfo->ContextRecord;
+    caughtExceptionAddress = pExceptionRecord->ExceptionAddress;
+    caughtInstructionPointer = pCONTEXT->Rip;
+    if (alreadyHandlingException == TRUE) {
+        return EXCEPTION_CONTINUE_SEARCH;
+    }
+    alreadyHandlingException = TRUE;
+    pCONTEXT->Rip = (uintptr_t)(&longjmpOutOfHere);
```

Comment:
Add some comments here.

> Basically, here, we coerce the os to resume us into a context that calls `longjmp` instead of just continuing. Presumably, we cannot `longjmp` out of the signal/exception context, like we can on unix.